README.md
# python_liquid_api_tths
[![Downloads](https://static.pepy.tech/personalized-badge/python-liquid-api-tths?period=total&units=international_system&left_color=black&right_color=orange&left_text=Downloads)](https://pepy.tech/project/python-liquid-api-tths)
## API概要
このライブラリはLiquid APIのPrivate APIとPublic APIをPythonで扱うものです。

PyPI
https://pypi.org/project/python-liquid-api-tths/

## インストール
インストールにはコンソールで次のコマンドを実行します。
```shell
pip install python-liquid-api-tths
```

## 目次
1. [Public API](#public)  
   1-1. [ローソク足(OHLCV)の生データを取得](#get_candlestick_raw)  
   1-2. [ローソク足(OHLCV)を取得](#get_candlestick)  
   1-3. [板情報の生データを取得](#get_order_book_raw)  
   1-4. [板情報の取得](#get_order_book)  
   1-5. [約定データの生データを取得](#get_execution_raw)  
   1-6. [約定データの取得](#get_execution)
2. [Private API](#private)  
   2-1. [注文](#order)  
   2-2. [注文のキャンセル](#order_cancel)  
   2-3. [資産残高の取得](#get_asset)  
   2-4. [日本円残高の取得（廃止予定メソッド）](#get_fiat)  
   2-5. [暗号資産残高の取得（廃止予定メソッド）](#get_crypto)  

## 1. <a id="public">Public API</a>

Public APIを使うにはLiquidPublicをインスタンス化します。引数は必要ありません。

```python
from python_liquid_api.public_api import LiquidPublic
pub = LiquidPublic()
```

### 1-1. <a id="get_candlestick_raw">ローソク足(OHLCV)の生データを取得</a>
ローソク足（OHLCVデータ）の生データを取得するにはget_candlestick_rawを使用します。
```python
from python_liquid_api.public_api import LiquidPublic
pub = LiquidPublic()
ohlc = pub.get_candlestick_raw(currency_name, candle_type)
```

#### 引数
- **currency_name**: 取得対象の通貨名を指定します。指定できる値は次の通りです。
  - btc: ビットコイン
  - eth: イーサリアム
  - xrp: リップル
  - bch: ビットコインキャッシュ
  - qash: キャッシュ
  - ltc: ライトコイン
  - bat: ベーシックアテンショントークン
- **candle_type**: 取得するローソク足のタイプを指定します。指定できる値は次の文字列です。
  - 1min: 1分足を取得
  - 5min: 5分足を取得
  - 15min: 15分足を取得
  - 30min: 30分足を取得
  - 1hour: 1時間足を取得


### 1-2. <a id="get_candlestick">ローソク足(OHLCV)を取得</a>
ローソク足（OHLCVデータ）を取得するにはget_candlestickを使用します。

```python
from python_liquid_api.public_api import LiquidPublic
pub = LiquidPublic()
ohlc = pub.get_candlestick(currency_name, date, candle_type)
```

#### 引数
- **currency_name**: 取得対象の通貨名を指定します。指定できる値は次の通りです。
  - btc: ビットコイン
  - eth: イーサリアム
  - xrp: リップル
  - bch: ビットコインキャッシュ
  - qash: キャッシュ
  - ltc: ライトコイン
  - bat: ベーシックアテンショントークン
- **date**: 取得対象の日付を指定します。フォーマットはyyyymmddの文字列です。
- **candle_type**: 取得するローソク足のタイプを指定します。指定できる値は次の文字列です。
  - 1min: 1分足を取得
  - 5min: 5分足を取得
  - 15min: 15分足を取得
  - 30min: 30分足を取得
  - 1hour: 1時間足を取得
- **is_index_datetime**: indexにdatetimeを設定する

#### 返り値
- **output_df**: pandas.DataFrame型のOHLCVが格納されたデータです。列は次の通りです。
  - datetime: datatime型の日時(yyyy-mm-dd hh:mm:ss)
  - open: 始値
  - high: 高値
  - low: 低値
  - close: 終値
  - volume: 出来高

#### 例外
- **通貨名が不正です。**: 引数のcurrency_nameに指定できる通貨名以外を指定した場合に発生します。
- **ローソク足のタイプが不正です。**: 引数のcandle_typeに指定できる値以外を指定した場合に発生します。


### 1-3. <a id="get_order_book_raw">板情報の生データを取得</a>
板情報を取得するにはget_order_book_rawを使用します。

```python
from python_liquid_api.public_api import LiquidPublic
pub = LiquidPublic()
raw_data = pub.get_order_book_raw(currency_name)
```

#### 引数
- **currency_name**: 取得対象の通貨名を指定します。指定できる値は次の通りです。
  - btc: ビットコイン
  - eth: イーサリアム
  - xrp: リップル
  - bch: ビットコインキャッシュ
  - qash: キャッシュ
  - ltc: ライトコイン
  - bat: ベーシックアテンショントークン

#### 返り値
- **raw_data**: 板情報

#### 例外
- **通貨名が不正です。**: 引数のcurrency_nameに指定できる通貨名以外を指定した場合に発生します。


### 1-4. <a id="get_order_book">板情報を取得</a>
板情報を取得するにはget_order_bookを使用します。

```python
from python_liquid_api.public_api import LiquidPublic
pub = LiquidPublic()
bid_df, ask_df, datetime_data = pub.get_order_book(currency_name)
```

#### 引数
- **currency_name**: 取得対象の通貨名を指定します。指定できる値は次の通りです。
  - btc: ビットコイン
  - eth: イーサリアム
  - xrp: リップル
  - bch: ビットコインキャッシュ
  - qash: キャッシュ
  - ltc: ライトコイン
  - bat: ベーシックアテンショントークン

#### 返り値
- **bid_df**: pandas.DataFrame型の売値データ
- **ask_df**: pandas.DataFrame型の買値データ
- **datetime_data**: datetime型の板情報取得日時

#### 例外
- **通貨名が不正です。**: 引数のcurrency_nameに指定できる通貨名以外を指定した場合に発生します。


### 1-5. <a id="get_execution_raw">約定情報の生データを取得</a>
約定情報を取得するにはget_executions_rawを使用します。

```python
from python_liquid_api.public_api import LiquidPublic
pub = LiquidPublic()
raw_data, url = pub.get_executions_raw(currency_name, timestamp, max_data_num=1000, base_url=None)
```

#### 引数
- **currency_name**: 取得対象の通貨名。指定できる値は次の通りです。
  - btc: ビットコイン
  - eth: イーサリアム
  - xrp: リップル
  - bch: ビットコインキャッシュ
  - qash: キャッシュ
  - ltc: ライトコイン
  - bat: ベーシックアテンショントークン
- **timestamp**: 取得対象のtimestamp
- **max_data_num**: 最大取得データ数
- **base_url**: 約定履歴を取得するURLのからtimestampを指定する部分を除いたURL。指定すると処理をいくつかスキップできます。

#### 返り値
- **raw_data**: 約定履歴の生データ
- **url**: 約定履歴を取得するURLのからtimestampを指定する部分を除いたURL


### 1-6. <a id="get_execution">約定情報を取得</a>
約定情報を取得するにはget_executionsを使用します。

```python
from python_liquid_api.public_api import LiquidPublic
pub = LiquidPublic()
execution_df = pub.get_executions(currency_name, date, hour)
```

#### 引数
- **currency_name**: 取得対象の通貨名。指定できる値は次の通りです。
  - btc: ビットコイン
  - eth: イーサリアム
  - xrp: リップル
  - bch: ビットコインキャッシュ
  - qash: キャッシュ
  - ltc: ライトコイン
  - bat: ベーシックアテンショントークン
- **date**: 取得対象の日付。フォーマットはyyyymmddの文字列
- **hour**: 取得対象の時間。フォーマットはhhの文字列

#### 返り値
- **out_df**: pandas.DataFrame型の約定データ
  - quantity: 取引量
  - price: 取引価格
  - taker_size: taker側のside(buy/sell)
  - timestamp: 取引された時刻


## 2. <a id="private">Private API</a>
Private APIを使うにはLiquidPrivateをインスタンス化します。

```python
from python_liquid_api.private_api import LiquidPrivate
pri = LiquidPrivate(token_id, secret_key)
```

#### 引数
- **token_id**: トークンID
- **secret_key**: APIトークン秘密鍵

### 2-1. <a id="order">注文</a>
現物取引の注文を出すにはcreate_orderを使用します。

```python
from python_liquid_api.private_api import LiquidPrivate
pri = LiquidPrivate(token_id, secret_key)
order_info = pri.create_order(currency_name, side, amount, price=0.0, order_type="limit")
```

#### 引数
- **currency_name**: 取得対象の通貨名。指定できる値は次の通りです。
  - btc: ビットコイン
  - eth: イーサリアム
  - xrp: リップル
  - bch: ビットコインキャッシュ
  - qash: キャッシュ
  - ltc: ライトコイン
  - bat: ベーシックアテンショントークン
- **side**: 売買の指定。指定できる値は次の通りです。
  - sell: 売り注文
  - buy: 買い注文
- **amount**: 注文量
- **price**: 注文単価
- **order_type**: 注文方法。指定できる値は次の通りです。
  - limit: 指値注文
  - market: 成行注文
  - stop: 逆指値注文

#### 返り値
- **output**: 辞書型の注文情報です。
  - transaction_id: 取引ID
  - order_type: 注文方法(limit/market)
  - quantity: 注文量
  - price: 取引単価
  - side: 売買(sell/buy)
  - created_at: 登録日時
  - updated_at: 更新日時
  - currency: 取引通貨名

#### 例外
- **通貨名が不正です。**: 引数のcurrency_nameに指定できる通貨名以外を指定した場合に発生します。
- **注文にはbuyもしくはsellを指定してください。**: 引数のsideに指定できる値以外を指定した場合に発生します。

### 2-2. <a id="order_cancel">注文のキャンセル</a>
注文をキャンセルするにはcancel_orderを使用します。

```python
from python_liquid_api.private_api import LiquidPrivate
pri = LiquidPrivate(token_id, secret_key)
pri.cancel_order(order_id)
```

#### 引数
- order_id: キャンセル対象の取引ID


### 2-3. <a id="get_asset">資産残高の取得</a>
資産残高の取得をするにはget_asset_infoを使用します。

```python
from python_liquid_api.private_api import LiquidPrivate
pri = LiquidPrivate(token_id, secret_key)
pri.get_asset_info(asset)
```
#### 引数
- **asset**:  取得対象資産
  - jpy: 日本円
  - btc: ビットコイン
  - eth: イーサリアム
  - xrp: リップル
  - bch: ビットコインキャッシュ
  - qash: キャッシュ
  - ltc: ライトコイン
  - bat: ベーシックアテンショントークン

#### 例外
- **通貨名が不正です。**: 引数のcurrency_nameに指定できる通貨名以外を指定した場合に発生します。


### 2-4. <a id="get_fiat">日本円残高の取得</a>
**! 廃止予定のメソッド get_asset_infoを使ってください。**

日本円残高を取得するにはget_fiat_infoを使用します。

```python
from python_liquid_api.private_api import LiquidPrivate
pri = LiquidPrivate(token_id, secret_key)
pri.get_fiat_info()
```
#### 返り値
- **balance**: 利用可能残高
- **reserved**: ロック中残高

#### 返り値
- **balance**: 利用可能残高
- **reserved**: ロック中残高



### 2-5. <a id="get_crypto">暗号資産残高の取得</a>
**! 廃止予定のメソッド get_asset_infoを使ってください。**
暗号資産残高を取得するにはget_crypto_infoを使用します。

```python
from python_liquid_api.private_api import LiquidPrivate
pri = LiquidPrivate(token_id, secret_key)
pri.get_crypto_info(currency="btc")
```

#### 引数
- **currency_name**: 取得対象の通貨名
  - btc: ビットコイン
  - eth: イーサリアム
  - xrp: リップル
  - bch: ビットコインキャッシュ
  - qash: キャッシュ
  - ltc: ライトコイン
  - bat: ベーシックアテンショントークン

#### 返り値
- **balance**: 利用可能残高
- **reserved**: ロック中残高


history.md
# 更新履歴
## 0.3.0
現物取引注文をするメソッド(create_order)で逆指値注文ができるようになりました。  
もともとorder_typeにstopを指定すれば逆指値注文することはできていましたが、動作確認をした上で引数に指定可能であることを明記しました。

## 0.3.2
注文をキャンセルするメソッド(cancel_order)を使用したときにエラーが出ていたので修正しました。

## 0.4.0
public APIにおいて生データを取得できるようにしました。
- get_candlestick_raw: ローソク足
- get_order_book_raw: 板情報
- get_executions_raw: 約定履歴

python_liquid_api/test_note.py
from python_liquid_api import public_api
from python_liquid_api import private_api
import os
import pandas as pd

pd.set_option("display.max_columns", None)
#a = public_api.LiquidPublic()
pri = private_api.LiquidPrivate(token_id=os.environ["id"], secret_key=os.environ["key"])
r = pri.create_order(currency_name="qash", side="buy", amount=1.0, order_type="market")
#r = pri.cancel_order("xxxxxxx)
#r = pri.get_order_info(limit_num=50)
#df = pd.DataFrame(r)

#print(df[df["filled_quantity"] == 0.0])

#balanced, reserved = pri.get_fiat_info()
#print(balanced, type(balanced))
#print(reserved, type(reserved))

balanced, reserved = pri.get_crypto_info(currency="BTC")
print(balanced, type(balanced))
print(reserved, type(reserved))

python_liquid_api/utils.py
# 定数
GOOD_CODE = "200"
UNKNOWN_ERROR_CONT = "不明なエラーです。"


def json_parse(request_result):
    """
    リクエスト結果のJSONデータをパースする。
    リクエストエラーがある場合は例外を起こす。
    """
    request_code = str(request_result.status_code)
    if request_code == GOOD_CODE:
        # JSONデータをパース
        return request_result.json()
    else:
        # エラーコードに対応する文章を取得
        if request_code in ERROR_CODES:
            error_contents = ERROR_CODES[request_code]
        else:
            error_contents = UNKNOWN_ERROR_CONT

        # エラーコードと対応する文章を出力
        message = "Error Code:" + request_code + " Contents:" + error_contents
        raise Exception(message)


def url_add_currency(url, currency_name):
    """
    URLに通貨IDを追加する
    """
    if currency_name in CURRENCY_ID:
        url += CURRENCY_ID[currency_name]
        return url
    else:
        raise ValueError("通貨名が不正です。")


def set_url(access_type, currency_name, resolution=None, max_data_num=1000):
    url = "https://api.liquid.com/"
    if access_type == "ohlc":
        # ローソク足
        # https://api.liquid.com/products/{product_id}/ohlc?resolution={resolution}
        url += "products/"
        url = url_add_currency(url, currency_name)
        url += "/ohlc?resolution="

        if resolution in CANDLE_TYPES:
            url += CANDLE_TYPES[resolution]
        else:
            raise ValueError("ローソク足のタイプが不正です。:", resolution)

    elif access_type == "book":
        # 板情報
        # https://api.liquid.com/products/{product_id}/price_levels?full=0
        url += "products/"
        url = url_add_currency(url, currency_name)
        url += "/price_levels?full=0"

    elif access_type == "executions":
        # 約定情報
        # https://api.liquid.com/executions?product_id={product_id}&limit={limit}&page={page}
        url += "executions" + "?product_id="
        url = url_add_currency(url, currency_name)
        url += "&limit=" + str(max_data_num)

    return url


# エラーコードと出力文章の対応
ERROR_CODES = {
    "404": "URLが存在しません。",
    "414": "URLが長すぎます。"
}

# 通貨略称と通貨IDの対応
CURRENCY_ID = {
    "btc": "5",
    "eth": "29",
    "xrp": "83",
    "bch": "41",
    "qash": "50",
    "ltc": "847",
    "bat": "846",
}

CANDLE_TYPES = {
    "1min": "60",
    "5min": "300",
    "15min": "900",
    "30min": "1800",
    "1hour": "3600",
}

python_liquid_api/private_api.py
import jwt
import json
import requests
from datetime import datetime
import warnings

from .parameter_dict import ParameterDict


class LiquidPrivate:
    def __init__(self, token_id, secret_key):
        """
        :param token_id:
        :param secret_key:
        """
        self.token_id = token_id
        self.secret_key = secret_key

        self.endpoint = "https://api.liquid.com/"
        self.parameter_dict = ParameterDict()

    def __make_header(self, path, query=""):
        timestamp = datetime.now().timestamp()
        path += query

        payload_data = {
            "path": path,
            "nonce": timestamp,
            "token_id": self.token_id
        }
        signature = jwt.encode(payload_data, self.secret_key, algorithm="HS256")
        header = {
            "X-Quoine-API-Version": "2",
            "X-Quoine-Auth": signature,
            "Content-Type": "application/json"
        }

        return path, header

    def __check_and_trans_params(self, currency_name, side, amount, price):
        currency_id = None

        if currency_name is not None:
            try:
                currency_id = self.parameter_dict.name2id[currency_name]
            except KeyError:
                raise ValueError("通貨名が不正です。")

        if side is not None:
            if side not in self.parameter_dict.side_list:
                raise ValueError("注文にはbuyもしくはsellを指定してください。")

        amount = float(amount)
        price = float(price)

        return currency_id, amount, price

    def create_order(self, currency_name, side, amount, price=0.0, order_type="limit"):
        """
        注文
        :param currency_name: 通貨名
        :param side: 売/買
        - sell: 売り注文
        - buy: 買い注文
        :param amount: 注文量
        :param price: 注文単価
        :param order_type: 注文タイプ
        - limit: 指値注文
        - market: 成行注文
        - stop: 逆指値
        :return:
        """
        url = self.endpoint + "orders/"
        currency_id, amount, price = self.__check_and_trans_params(
            currency_name=currency_name,
            side=side,
            amount=amount,
            price=price
        )

        # ヘッダ情報作成
        url, header = self.__make_header(path=url)

        # 注文データ
        send_data = {
            "order": {
                "order_type": order_type,
                "product_id": currency_id,
                "side": side,
                "quantity": amount,
                "price": price,
            }
        }

        # 送信データ作成
        json_data = json.dumps(send_data)
        res = requests.post(url=url, headers=header, data=json_data)

        parsed_data = json.loads(res.text)
        create_datetime = datetime.fromtimestamp(parsed_data["created_at"])
        create_datetime = datetime.strftime(create_datetime, "%Y/%m/%d %H:%M:%S")

        update_datetime = datetime.fromtimestamp(parsed_data["updated_at"])
        update_datetime = datetime.strftime(update_datetime, "%Y/%m/%d %H:%M:%S")

        output = {
            "transaction_id": str(parsed_data["id"]),
            "order_type": parsed_data["order_type"],
            "quantity": float(parsed_data["quantity"]),
            "price": float(parsed_data["price"]),
            "side": parsed_data["side"],
            "created_at": create_datetime,
            "updated_at": update_datetime,
            "currency": parsed_data["currency_pair_code"][:-3]  # 取引通貨のみ取得（文字列からJPYを除く）
        }

        return output

    def get_order_info(self, limit_num=None):
        url = self.endpoint + "orders"
        query = None

        # 取得件数を指定した場合はクエリを設定
        if limit_num is not None:
            query = "?limit=" + str(limit_num)

        # ヘッダ情報作成
        url, header = self.__make_header(path=url, query=query)
        # データ送信
        res = requests.get(url=url, headers=header)

        if res.status_code != 200:
            print("注文情報の取得に失敗しました。")
            return None

        parsed_data = json.loads(res.text)["models"]
        output_list = []
        for data in parsed_data:
            create_datetime = datetime.fromtimestamp(data["created_at"])
            create_datetime = datetime.strftime(create_datetime, "%Y/%m/%d %H:%M:%S")

            update_datetime = datetime.fromtimestamp(data["updated_at"])
            update_datetime = datetime.strftime(update_datetime, "%Y/%m/%d %H:%M:%S")

            # 出力データ作成
            output_dict = {
                "transaction_id": str(data["id"]),
                "order_type": data["order_type"],
                "price": float(data["price"]),
                "quantity": float(data["quantity"]),
                "filled_quantity": float(data["filled_quantity"]),  # 約定済の量
                "side": data["side"],
                "created_at": create_datetime,
                "updated_at": update_datetime,
                "currency": data["currency_pair_code"][:-3]  # 取引通貨のみ取得（文字列からJPYを除く）
            }

            output_list.append(output_dict)

        return output_list

    def cancel_order(self, order_id):
        url = self.endpoint + "orders/" + order_id + "/cancel"
        # ヘッダ情報作成
        path, header = self.__make_header(path=url)
        # データ送信
        res = requests.put(url=url, headers=header)

        if res.status_code == 404:
            print("対象の取引IDが存在しません。取引ID:", order_id)
        elif res.status_code == 200:
            print("注文がキャンセルされました。取引ID:", order_id)

    def get_fiat_info(self):
        """
        日本円残高取得
        :return:
        - balanced: 利用可能残高
        - reserved: ロック中残高
        """
        warnings.warn("廃止予定のメソッドです。get_asset_info(asset='jpy')を使用してください。", category=FutureWarning)
        url = self.endpoint + "fiat_accounts"
        # ヘッダ情報作成
        url, header = self.__make_header(path=url)
        # データ送信
        res = requests.get(url=url, headers=header)
        parsed = json.loads(res.text)[0]

        balance = parsed["balance"]  # 日本円残高
        reserved = parsed["reserved_balance"]  # ロック中
        return balance, reserved

    def get_crypto_info(self, currency="btc"):
        warnings.warn("廃止予定のメソッドです。get_asset_info(asset)を使用してください。", category=FutureWarning)
        url = self.endpoint + "crypto_accounts"
        # ヘッダ情報作成
        url, header = self.__make_header(path=url)
        # データ送信
        res = requests.get(url=url, headers=header)
        parsed = json.loads(res.text)

        currency = currency.upper()
        data = None
        # 引数で与えた通貨の情報を取得
        for p in parsed:
            if p["currency"] == currency:
                data = p
                break

        balance = data["balance"]  # 利用可能残高
        reserved = data["reserved_balance"]  # ロック中残高

        return balance, reserved

    def get_asset_info(self, asset):
        url = self.endpoint
        # 日本円を指定した場合
        if asset in self.parameter_dict.fiat_list:
            url += "fiat_accounts"
        # 暗号資産を指定した場合
        elif asset in self.parameter_dict.name2id:
            url += "crypto_accounts"
        else:
            raise Exception("通貨名が不正です。")

        # ヘッダ情報作成
        url, header = self.__make_header(path=url)
        # データ送信
        res = requests.get(url=url, headers=header)
        parsed = json.loads(res.text)
        asset = asset.upper()

        data = None
        # 引数で与えた通貨の情報を取得
        for p in parsed:
            if p["currency"] == asset:
                data = p
                break

        balance = data["balance"]  # 利用可能残高
        reserved = data["reserved_balance"]  # ロック中残高

        return balance, reserved


python_liquid_api/public_api.py
from .utils import json_parse, url_add_currency, set_url
import requests
import json
import datetime
import pandas as pd
import warnings


class LiquidPublic(object):
    @staticmethod
    def get_candlestick_raw(currency_name, candle_type):
        """
        ローソク足を取得して生データを出力
        :param currency_name: 通貨名
        :param candle_type: ローソク足範囲
        - 1min
        - 5min
        - 15min
        - 30min
        - 1hour
        :return: ローソク足Data
        - datetime: UNIX Timestamp
        - open
        - high
        - low
        - close
        - volume
        """
        # URLの設定
        set_url_params = {
            "access_type": "ohlc",
            "currency_name": currency_name,
            "resolution": candle_type
        }
        url = set_url(**set_url_params)

        # APIからローソク足を取得
        req_result = requests.get(url)
        raw_data = json_parse(req_result)["data"]

        return raw_data

    def get_candlestick(self, currency_name, date, candle_type, is_index_datetime=True):
        """
        ローソク足を取得
        :param currency_name: 通貨名
        :param date: 取得対象日付
        :param candle_type: ローソク足範囲
        - 1min
        - 5min
        - 15min
        - 30min
        - 1hour
        :param is_index_datetime: indexにdatetimeを設定
        :return: ローソク足DataFrame
        - datetime: yyyy-mm-dd hh:mm:ss
        - open
        - high
        - low
        - close
        - volume
        """
        # ローソク足の生データを取得
        parsed_data = self.get_candlestick_raw(currency_name, candle_type)

        # DataFrameに変換
        output_df = pd.DataFrame(data=parsed_data, columns=["datetime", "open", "high", "low", "close", "volume"])
        output_df["datetime"] = output_df["datetime"].apply(datetime.datetime.fromtimestamp)

        # 引数のdateをdatetime型に変換
        target_date = datetime.datetime.strptime(date, "%Y%m%d")
        target_date_next = target_date + datetime.timedelta(days=1)

        # 引数で指定したデータを取得
        output_df = output_df.loc[(output_df["datetime"] >= target_date) & (output_df["datetime"] < target_date_next)]
        output_df = output_df.reset_index(drop=True)

        if is_index_datetime:
            # indexにdatetimeを設定
            output_df.index = output_df["datetime"]
            # datetime列を削除
            output_df = output_df.drop("datetime", axis=1)

        if len(output_df) == 0:
            warnings.warn("指定日付が取得範囲外です。")

        return output_df

    @staticmethod
    def get_order_book_raw(currency_name):
        """
        板情報の生データを取得
        - buy_price_levels: 買値
        - sell_price_levels: 売値
        - timestamp: 取得日時
        """
        # URLの設定
        set_url_params = {
            "access_type": "book",
            "currency_name": currency_name,
        }
        url = set_url(**set_url_params)

        # 板情報の生データを取得
        req_result = requests.get(url)
        raw_data = json_parse(req_result)

        return raw_data

    def get_order_book(self, currency_name):
        """
        板情報を取得
        :param currency_name: 通貨名
        :return: 売値DataFrame, 買値DataFrame, datetime
        """
        # 板情報の生データを取得
        order_book_raw = self.get_order_book_raw(currency_name)

        # 売値
        bid_df = pd.DataFrame(
            data=order_book_raw["sell_price_levels"],
            columns=["bid_price", "bid_volume"]
        )

        # 買値
        ask_df = pd.DataFrame(
            data=order_book_raw["buy_price_levels"],
            columns=["ask_price", "ask_volume"]
        )

        # 板情報取得日時
        datetime_data = datetime.datetime.fromtimestamp(float(order_book_raw["timestamp"]))
        return bid_df, ask_df, datetime_data

    @staticmethod
    def get_executions_raw(currency_name, timestamp, max_data_num=1000, base_url=None):
        """
        約定履歴を取得
        :param currency_name: 通貨名
        :param timestamp: 取得対象timestamp
        :param max_data_num: 最大取得データ数
        :param base_url: 約定履歴を取得するURLのからtimestampを指定する部分を除いたURL
        """
        if base_url is None:
            # URLを設定
            set_url_params = {
                "access_type": "executions",
                "currency_name": currency_name,
                "max_data_num": max_data_num,
            }
            url_tmp = set_url(**set_url_params)
        else:
            url_tmp = base_url

        url = url_tmp + "&timestamp=" + str(timestamp)
        req_result = requests.get(url)
        raw_data = json_parse(req_result)
        return raw_data, url_tmp

    def get_executions(self, currency_name, date, hour):
        """
        約定履歴を取得
        :param currency_name: 通貨名
        :param date: 日付(yyyymmdd)
        :param hour: 時間(hh)
        """
        def timestamp2datetime(timestamp):
            return datetime.datetime.fromtimestamp(float(timestamp))

        # 取得対象のtimestampを取得
        target_timestamp = datetime.datetime.strptime(date+hour, "%Y%m%d%H").timestamp()
        end_timestamp = target_timestamp + 60*60  # 開始時点のtimestampの1時間(3600秒)後

        # 出力用DataFrame
        out_df = pd.DataFrame(columns=["id", "quantity", "price", "taker_side", "created_at", "timestamp"])

        url = None
        # 比較対象のtimestampが開始時点のtimestampの1時間(3600秒)後までLOOP
        while target_timestamp < end_timestamp:
            # 約定生データと基底URLを取得
            raw_data, url = self.get_executions_raw(currency_name, target_timestamp, base_url=url)

            # 約定データが取得できなかった場合
            if len(raw_data) == 0:
                if len(out_df) == 0:
                    # 出力用DataFrameがまだ作成されていない場合は処理を終了
                    return pd.DataFrame()
                else:
                    # 出力用DataFrameが作成されている場合は加工処理へ
                    break

            # DataFrameを作成
            df = pd.DataFrame(raw_data)
            merge_df = pd.merge(out_df, df, on="timestamp")

            # 約定データがすべて取得済の場合
            if len(merge_df) == len(df):
                if len(out_df) == 0:
                    # 出力用DataFrameがまだ作成されていない場合は処理を終了
                    return pd.DataFrame()
                else:
                    # 出力用DataFrameが作成されている場合は加工処理へ
                    break

            target_timestamp = df.tail(1)["timestamp"].values[0]  # 最後のレコードのtimestampを次の取得地点とする
            target_timestamp = float(target_timestamp)  # whileで比較するためにfloatに変換

            out_df = pd.concat([out_df, df], axis=0)

        # 出力用に加工
        out_df = out_df.loc[out_df["timestamp"] < str(end_timestamp)]  # 終了timestampより前のレコードのみ取得
        out_df["timestamp"] = out_df["timestamp"].apply(timestamp2datetime)  # UNIX時間を変換
        out_df = out_df.drop_duplicates(subset="id")  # idで重複削除
        out_df = out_df.reset_index(drop=True)  # indexを削除
        out_df = out_df.drop(["created_at", "id"], axis=1)  # 不要な列を削除

        return out_df
